### PR TITLE
Add draft for ERC-3772

### PR DESCRIPTION
**Pull request by zemse**
This PR adds a draft of #3772. Pls let me know if any changes are needed.

### CODE REVIEW

File: EIPS/eip-3772.md [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+discussions-to: https://github.com/ethereum/EIPs/issues/3772
+status: Draft
+type: Standards Track
+category: ERC
+created: 2021-08-27
+---
+
+## Abstract
+
+This document specifies compression of `uint256` to smaller data structures like `uint64`, `uint96`, `uint128`, for optimizing costs for storage in Ethereum Smart Contracts. The smaller data structure (represented as `cintx`) is divided into two parts, in the first one we store `significant` bits and in the other number of left `shift`s needed on the significant bits to decompress. This document also includes two specifications for decompression due to the nature of compression being lossy, i.e. it causes underflow.
```
Comment (lightclient) [inline EIPS/eip-3772.md:15]:
```suggestion
This document specifies compression of `uint256` to smaller data structures like `uint64`, `uint96`, `uint128`, for optimizing costs for storage. The smaller data structure (represented as `cintx`) is divided into two parts, in the first one we store `significant` bits and in the other number of left `shift`s needed on the significant bits to decompress. This document also includes two specifications for decompression due to the nature of compression being lossy, i.e. it causes underflow.
```

Comment (lightclient) [reply EIPS/eip-3772.md:15]: Ethereum is assumed :).

File: EIPS/eip-3772.md [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+
+## Motivation
+
+- Storage is costly, each storage slot costs almost $0.8 to initialize and $0.2 to update (20 gwei, 2000 ETHUSD).
+- Usually, we store money amounts in uint256 which takes up one entire slot.
```
Comment (lightclient) [inline EIPS/eip-3772.md:20]:
```suggestion
- Usually, we store money amounts in `uint256` which takes up one entire slot.
```

File: EIPS/eip-3772.md [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+- If it's DAI value, the range we work with most is 0.001 DAI to 1T DAI (or 10<sup>12</sup>). If it's ETH value, the range we work with most is 0.000001 ETH to 1B ETH. Similarly, any token of any scale has a reasonable range of 10<sup>15</sup> amounts that we care/work with.
+- However, uint256 type allows us to represent $10<sup>-18</sup> to $10<sup>58</sup>, and most of it is a waste. In technical terms, we have the probability distribution for values larger than $10<sup>15</sup> and smaller than $10<sup>-3</sup> as negligible (i.e. P[val > 10<sup>15</sup>] ≈ 0 and P[val < 10<sup>-3</sup>] ≈ 0).
+- Number of bits required to represent 10<sup>15</sup> values = log<sub>2</sub>(10<sup>15</sup>) = 50 bits. So just 50 bits (instead of 256) are reasonably enough to represent a practical range of money, causing a very negligible difference.
+
+## Specification
+
+In this specification, the structure for representing a compressed value is represented using `cintx`, where x is the number of bits taken by the entire compressed value. (On the implementation level, an `uintx` can be used for storing a `cintx` value).
```
Comment (lightclient) [inline EIPS/eip-3772.md:27]:
```suggestion
In this specification, the structure for representing a compressed value is represented using `cintx`, where x is the number of bits taken by the entire compressed value. On the implementation level, an `uintx` can be used for storing a `cintx` value.
```

File: EIPS/eip-3772.md [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+
+### Compression
+
+#### uint256 into cint64 (up to cint120)
+
+The rightmost or least significant 8 bits in `cintx` are reserved for storing the shift and the rest available bits are used to store the significant bits starting from the first `1` bit in `uintx`.
```
Comment (lightclient) [inline EIPS/eip-3772.md:33]:
```suggestion
The rightmost, or least significant, 8 bits in `cintx` are reserved for storing the shift and the rest available bits are used to store the significant bits starting from the first `1` bit in `uintx`.
```

File: EIPS/eip-3772.md [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+struct cint64 { uint56 significant; uint8 shift; }
+
+// ...
+
+struct cint120 { uint112 significant; uint8 shift; }
+```
+
+#### uint256 into cint128 (up to cint248)
+
+The rightmost or least significant 7 bits in `cintx` are reserved for storing the shift and the rest available bits are used to store the significant bits starting from the first on bit in `uintx`.
```
Comment (lightclient) [inline EIPS/eip-3772.md:45]:
```suggestion
The rightmost, or least significant, 7 bits in `cintx` are reserved for storing the shift and the rest available bits are used to store the significant bits starting from the first one bit in `uintx`.
```

Comment (lightclient) [reply EIPS/eip-3772.md:45]: `s/on/one` typo?

File: EIPS/eip-3772.md [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+    address beneficiary;
+    bytes32 hash;
+}
+```
+
+## Backwards Compatibility
+
+There are no known backward-incompatible issues.
+
+<!-- ## Test Cases
```
Comment (lightclient) [inline EIPS/eip-3772.md:146]: Please remove these comments.

File: EIPS/eip-3772.md [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+
+Test cases for an implementation are mandatory for EIPs that are affecting consensus changes. If the test suite is too large to reasonably be included inline, then consider adding it as one or more files in `../assets/eip-####/`. -->
+
+## Reference Implementation
+
+<!-- An optional section that contains a reference/example implementation that people can use to assist in understanding or implementing this specification. If the implementation is too large to reasonably be included inline, then consider adding it as one or more files in `../assets/eip-####/`. -->
```
Comment (lightclient) [inline EIPS/eip-3772.md:152]: Please remove.

**Review by lightclient — CHANGES_REQUESTED** [review]
Nice work on this EIP. I've left a few comments - thanks.

**Comment by zemse** [comment]
@lightclient Thanks for your time in reviewing this. I have incorporated the changes.

File: EIPS/eip-3772.md [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+
+### Some related work
+
+Some engineers have realized that using `uint256` entirely for storing value/money is not optimal in view of storage costs, and some of the bits in the slot could be better utilized with other data stored in its place.
+
+- Sushiswap's MasterChefV2: ([code ref](https://github.com/sushiswap/sushiswap/blob/master/contracts/MasterChefV2.sol#L198), they are casting `uint256` values to `uint128`, to make some room for other data. No lossy compression).
+- Dydx's Balance ([code ref](https://github.com/dydxprotocol/perpetual/blob/13eea6201cd5a89814d1f90aae2382456289c2f3/contracts/protocol/v1/lib/P1Types.sol#L45-L50), uses `uint120`, no lossy compression).
+- Compound's GovernorAlpha ([code ref](https://github.com/compound-finance/compound-protocol/blob/master/contracts/Governance/GovernorAlpha.sol#L88), they use `uint96`)
+- Rollups ([Using about 3 bytes or 24 bits for representing ETH value](https://vitalik.ca/general/2021/01/05/rollup.html#how-does-compression-work))
+
```
Comment (MicahZoltu) [inline EIPS/eip-3772.md:184]: EIPs are not the appropriate place for this kind of history lesson, and should not contain external links.

**Pull request merged**